### PR TITLE
Laboratory work #1, Eliseykina Kate - 20FPL1

### PR DESCRIPTION
**Pull request by reginafeles**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
It's time to push some changes. October 1st is coming

File: lab_1/main.py [hunk 1]
```diff
@@ -3,81 +3,148 @@
...
+def tokenize(text: str):
     """
     Splits a text into tokens, converts the tokens into lowercase,
     removes punctuation and other symbols from words
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
-    """
-    pass
+        """
+    if isinstance(text, str):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:16]: let's put checks with None at the beginning
this way the code would be more readable

File: lab_1/main.py [hunk 2]
```diff
@@ -3,81 +3,148 @@
...
-def remove_stop_words(tokens: list, stop_words: list) -> list or None:
+def remove_stop_words(tokens: list, stop_words: list):
     """
     Removes stop words
     :param tokens: a list of tokens
     :param stop_words: a list of stop words
     :return: a list of tokens without stop words
     """
-    pass
+    try:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:34]: wow, why do you need `try/except`? maybe a simple `if` will do the job?

File: lab_1/main.py [hunk 3]
```diff
@@ -3,81 +3,148 @@
...
+        for i in tokens:
+            if not isinstance(i, str):
+                return None
+            tokens = [i for i in tokens if i not in stop_words]
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:38]: You iterate over tokens and then on EACH iteration you create the same list...

File: lab_1/main.py (hunk 2 above)
Comment (ValeriyaKuznetsova) [reply lab_1/main.py:34]: and as I said, checks with None to the beginning

File: lab_1/main.py [hunk 4]
```diff
@@ -3,81 +3,148 @@
...
+            return tokens
+    except TypeError:
+        return None
 
 
-def calculate_frequencies(tokens: list) -> dict or None:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:27]: Why did you remove `-> dict or None`?
these lines are quite useful

change back everywhere

File: lab_1/main.py [hunk 5]
```diff
@@ -3,81 +3,148 @@
...
+def calculate_frequencies(tokens: list):
     """
     Calculates frequencies of given tokens
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
-    pass
+    if isinstance(tokens, list):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:50]: check

File: lab_1/main.py [hunk 6]
```diff
@@ -3,81 +3,148 @@
...
+        for i in tokens:
+            if isinstance(i, str):
+                freq_dict = {i: tokens.count(i) for i in tokens}
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:53]: don't use generators in a loop

poor computer...

File: lab_1/main.py [hunk 7]
```diff
@@ -3,81 +3,148 @@
...
     Returns the most common words
     :param freq_dict: a dictionary with frequencies
     :param top_n: a number of the most common words
     :return: a list of the most common words
     """
-    pass
-
-
-def create_language_profile(language: str, text: str, stop_words: list) -> dict or None:
+    if isinstance(freq_dict, dict) and isinstance(top_n, int):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:65]: check

File: lab_1/main.py [hunk 8]
```diff
@@ -3,81 +3,148 @@
...
+        if freq_dict != {} and top_n > 0:
+            new_dict = sorted(freq_dict.items(), key=lambda x: x[1], reverse=True)
+            count = 1
+            top_list = []
+            first, snd = zip(*new_dict)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:70]: wow zip

File: lab_1/main.py [hunk 9]
```diff
@@ -3,81 +3,148 @@
...
+            del snd
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:71]: wow del snd

File: lab_1/main.py [hunk 10]
```diff
@@ -3,81 +3,148 @@
...
+            for i in first:
+                if count <= top_n:
+                    top_list.append(i)
+                    count += 1
+                if count > top_n:
+                    break
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:77]: don't get this loop
maybe you can sort keys, not items, here? `new_dict = sorted(freq_dict.items(), key=lambda x: x[1], reverse=True)`

File: lab_1/main.py [hunk 11]
```diff
@@ -3,81 +3,148 @@
...
     :param language: a language
     :param text: a text
     :param stop_words: a list of stop words
     :return: a dictionary with three keys – name, freq, n_words
     """
-    pass
+    first_step = tokenize(text)
+    second_step = remove_stop_words(first_step, stop_words)
+    freq_dict = calculate_frequencies(second_step)
+    if isinstance(language, str) and isinstance(text, str) and isinstance(stop_words, list):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:94]: check at the beginning

File: lab_1/main.py [hunk 12]
```diff
@@ -3,81 +3,148 @@
...
-    :param profile_1: a dictionary
-    :param profile_2: a dictionary
-    :param top_n: a number of the most common words
-    :return: a language
-    """
-    pass
-
-
-def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, top_n: int) -> list or None:
+    if isinstance(unknown_profile, dict) \
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:110]: check

File: lab_1/main.py [hunk 13]
```diff
@@ -3,81 +3,148 @@
...
+            and isinstance(profile_to_compare, dict) \
+            and isinstance(top_n, int):
+        top_n_common_words = []
+        top_n_words_unknown_profile = get_top_n_words(unknown_profile['freq'], top_n)
+        top_n_words_known_profile = get_top_n_words(profile_to_compare['freq'], top_n)
+        for i in top_n_words_unknown_profile:
+            if i in top_n_words_known_profile:
+                top_n_common_words.append(i)
+        intersecting_words = len(top_n_common_words) / len(top_n_words_unknown_profile)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:119]: better to name it `distance` or smth

File: lab_1/main.py [hunk 14]
```diff
@@ -3,81 +3,148 @@
...
+            and isinstance(profile_1, dict) \
+            and isinstance(profile_2, dict) \
+            and isinstance(top_n, int):
+        first_intersecting_words = compare_profiles(unknown_profile, profile_1, top_n)
+        second_intersecting_words = compare_profiles(unknown_profile, profile_2, top_n)
+        if first_intersecting_words > second_intersecting_words:
+            return profile_1['name']
+        if second_intersecting_words > first_intersecting_words:
+            return profile_2['name']
+        return [[profile_1['name'], profile_2['name']].sort()][0]
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:143]: can you describe in writing what happens here?

File: lab_1/main.py [hunk 15]
```diff
@@ -3,81 +3,148 @@
...
+def detect_language(unknown_profile: dict, profile_1: dict, profile_2: dict, top_n: int):
+    """
+        Detects the language of an unknown profile
+        :param unknown_profile: a dictionary
+        :param profile_1: a dictionary
+        :param profile_2: a dictionary
+        :param top_n: a number of the most common words
+        :return: a language
+        """
+    if isinstance(unknown_profile, dict) \
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:133]: check

File: lab_1/main.py [hunk 16]
```diff
@@ -86,34 +153,98 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
     :return: a dictionary with 7 keys – name, score, common, sorted_common, max_length_word,
     min_length_word, average_token_length
     """
-    pass
-
-
-def detect_language_advanced(unknown_profile: dict, profiles: list, languages: list, top_n: int) -> str or None:
+    if isinstance(unknown_profile, dict) \
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:156]: check

File: lab_1/main.py [hunk 17]
```diff
@@ -86,34 +153,98 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
...
+            and isinstance(profile_to_compare, dict) \
+            and isinstance(top_n, int):
+        common = []
+        top_n_words_unknown_profile = get_top_n_words(unknown_profile['freq'], top_n)
+        top_n_words_known_profile = get_top_n_words(profile_to_compare['freq'], top_n)
+        for i in top_n_words_known_profile:
+            if i in top_n_words_unknown_profile:
+                common.append(i)
+        sorted_common = sorted(common)
+        score = (len(common) / len(top_n_words_known_profile))
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:166]: brackets?

File: lab_1/main.py [hunk 18]
```diff
@@ -86,34 +153,98 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
...
+        profile_to_compare_keys = list(profile_to_compare['freq'].keys())
+        sorted_keys = sorted(profile_to_compare_keys, key=len)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:168]: can you explain what happens here? In writing

File: lab_1/main.py [hunk 19]
```diff
@@ -86,34 +153,98 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
...
+    :param profiles: a dictionary
     :param languages: a list of possible languages
     :param top_n: a number of the most common words
     :return: a language
     """
-    pass
-
-
-def load_profile(path_to_file: str) -> dict or None:
+    if isinstance(unknown_profile, dict) \
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:192]: check

File: lab_1/main.py [hunk 20]
```diff
@@ -86,34 +153,98 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
...
+            and isinstance(profiles, list) \
+            and isinstance(languages, list) \
+            and isinstance(top_n, int):
+        def score_function():
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:196]: a function in a function  – no, either put it outside a function or delete def

File: lab_1/main.py [hunk 21]
```diff
@@ -86,34 +153,98 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
...
+        for i in top_n_words_known_profile:
+            if i in top_n_words_unknown_profile:
+                common.append(i)
+        sorted_common = sorted(common)
+        score = (len(common) / len(top_n_words_known_profile))
+        profile_to_compare_keys = list(profile_to_compare['freq'].keys())
+        sorted_keys = sorted(profile_to_compare_keys, key=len)
+        max_length_word = sorted_keys[-1]
+        min_length_word = sorted_keys[0]
+        average_token_length = len(''.join(profile_to_compare_keys)) / len(profile_to_compare_keys)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:171]: not bad

File: lab_1/main.py [hunk 22]
```diff
@@ -86,34 +153,98 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
...
+            and isinstance(top_n, int):
+        def score_function():
+            score = {}
+            for language in profiles:
+                if language['name'] in languages:
+                    full_profile = compare_profiles_advanced(unknown_profile, language, top_n)
+                    score.update({language['name']: full_profile['score']})
+            if not score:
+                return None
+            score_sorted = sorted(score.items(), key=lambda x: x[1], reverse=True)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:204]: sort keys directly

File: lab_1/main.py [hunk 23]
```diff
@@ -86,34 +153,98 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
...
+            return [a[0] for a in score_sorted][0]
+
+        if languages:
+            max_score = score_function()
+            return max_score
+        if not languages:
+            for i in profiles:
+                languages.append(i['name'])
+            max_score = score_function()
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:213]: duplicate code? maybe you don't need two ifs then
```
max_score = score_function 
return max_score
```

File: lab_1/main.py [hunk 24]
```diff
@@ -86,34 +153,98 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
...
+
+
+def load_profile(path_to_file: str):
     """
     Loads a language profile
     :param path_to_file: a path
     :return: a dictionary with three keys – name, freq, n_words
     """
-    pass
+    if isinstance(path_to_file, str):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:224]: why if and then try?

Comment (ValeriyaKuznetsova) [inline lab_1/main.py:224]: check

File: lab_1/main.py [hunk 25]
```diff
@@ -86,34 +153,98 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
...
 
-def save_profile(profile: dict) -> int:
+def save_profile(profile: dict):
     """
     Saves a language profile
     :param profile: a dictionary
     :return: 0 if everything is ok, 1 if not
     """
-    pass
+    if isinstance(profile, dict):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:240]: check

File: lab_1/profiles/unk.json [hunk 26]
```diff
@@ -0,0 +1 @@
...", "freq": {"at": 1, "first": 1, "von": 1, "frisch": 1, "thought": 1, "the": 19, "bees": 3, "were": 2, "responding": 2, "only": 2, "to": 4, "scent": 2, "of": 5, "food": 3, "but": 2, "what": 1, "did": 2, "third": 2, "dance": 3, "mean": 1, "and": 3, "if": 2, "how": 1, "could": 1, "they": 1, "also": 2, "sniff": 1, "down": 1, "hundreds": 1, "metres": 5, "away": 5, "from": 1, "hive": 1, "which": 1, "was": 4, "sometimes": 1, "downwind": 1, "on": 1, "a": 2, "hunch": 1, "he": 5, "started": 2, "gradually": 1, "moving": 1, "feeding": 3, "dish": 4, "further": 2, "noticed": 1, "as": 1, "so": 1, "that": 1, "dances": 1, "returning": 1, "scout": 1, "changing": 1, "placed": 1, "over": 1, "nine": 1, "second": 1, "type": 1, "sickle": 1, "version": 1, "came": 1, "into": 1, "play": 1, "once": 1, "moved": 2, "it": 1, "past": 1, "scouts": 1, "would": 1, "then": 1, "start": 1, "dancing": 1, "quite": 1, "different": 1, "waggle": 1, "measurement": 1, "actual": 1, "distance": 1, "too": 1, "concluded": 1, "precise": 1, "for": 1, "example": 1, "indicated": 1, "by": 1, "complete": 1, "runs": 1, "through": 1, "pattern": 1, "in": 1, "seconds": 1, "when": 1, "number": 1, "dropped": 1, "eleven": 1}, "n_words": 97}
```
Comment (ValeriyaKuznetsova) [inline lab_1/profiles/unk.json:1]: remove this file

File: lab_1/start.py [hunk 27]
```diff
@@ -15,11 +16,31 @@
...
+    main.save_profile(unk)
+
+    de_profile = main.load_profile(os.path.join(PATH_TO_LAB_FOLDER, r'profiles\de.json'))
+    en_profile = main.load_profile(os.path.join(PATH_TO_LAB_FOLDER, r'profiles\en.json'))
+    la_profile = main.load_profile(os.path.join(PATH_TO_LAB_FOLDER, r'profiles\la.json'))
+    unk_profile = main.load_profile(os.path.join(PATH_TO_LAB_FOLDER, r'profiles\unk.json'))
+
+    languages = ['en', 'de', 'la']
+    profiles = [en_profile, de_profile, la_profile]
+    RESULT = main.detect_language_advanced(unk_profile,profiles ,languages , 5)
```
Comment (ValeriyaKuznetsova) [inline lab_1/start.py:39]: can you describe in writing what happens in `start.py`?

**Review by ValeriyaKuznetsova — COMMENTED** [review]
Don't be afraid of the number of comments :)
Your lab is good but some things can be improved.

I have some questions. They are obligatory to answer – here or via Telegram

File: lab_1/main.py (hunk 14 above)
Comment (reginafeles) [reply lab_1/main.py:143]: first_intersecting_words and second_intersecting_words-float(proportion of intersecting words of the profile_1 and profile_2 that is returned by  compare_profiles) 
with their help we can detect the language:
-if the proportion of the language from first profile will be greater than from the second, the name of the first language returns and that will be the name of the unknown language
-however, if proportions are the same, names sorted alphabetically and the unkown language will be the language with index 0

File: lab_1/main.py (hunk 1 above)
Comment (reginafeles) [reply lab_1/main.py:16]: done

File: lab_1/main.py (hunk 2 above)
Comment (reginafeles) [reply lab_1/main.py:34]: Yes, it works now without try/exept!
done

File: lab_1/main.py (hunk 3 above)
Comment (reginafeles) [reply lab_1/main.py:38]: omg ... i have just realised it. thanks. i improved it

File: lab_1/main.py (hunk 13 above)
Comment (reginafeles) [reply lab_1/main.py:119]: done

File: lab_1/main.py (hunk 17 above)
Comment (reginafeles) [reply lab_1/main.py:166]: removed

File: lab_1/main.py (hunk 20 above)
Comment (reginafeles) [reply lab_1/main.py:196]: decided to convert it in if

File: lab_1/main.py (hunk 22 above)
Comment (reginafeles) [reply lab_1/main.py:204]: done

File: lab_1/main.py (hunk 24 above)
Comment (reginafeles) [reply lab_1/main.py:224]: because without exceptions code did not work. now i changed the idea a little bit

File: lab_1/profiles/unk.json (hunk 26 above)
Comment (reginafeles) [reply lab_1/profiles/unk.json]: done

File: lab_1/start.py (hunk 27 above)
Comment (reginafeles) [reply lab_1/start.py:39]: import function from main
read text from files that we have to variables in type string
the main aim is to detect unkown language using profiles(*.json) and texts.
we have 2 ways how we can do it:
-by our own
-using external  profiles
with the help of texts we create  language profiles(function create_language_profile) and then detect language(function detect_language_advanced). this is result numer 1
then we load external profiles(function load_profile) and again detect(function detect_language_advanced).this is result number 2
and finally we compare results: if they are identical, everything works great , however, if there is 'Detection not working' something doesn't work)))
that's all

File: lab_1/main.py (hunk 24 above)
Comment (reginafeles) [reply lab_1/main.py]: and everything is ok

File: lab_1/main.py (hunk 18 above)
Comment (reginafeles) [reply lab_1/main.py]: first of all, we get **top n words of unknown and known language** using function get_top_n_words. then we use these variables to get list of **common words** in cycle. then we divide number of common words by nuber of top n words of unknown language and in the end we get **score** (proportion of  intersecting words). then the length of the **shortest word and the longest** is counted here(profile_to_compare_keys = list(profile_to_compare['freq'].keys())- pull words out of the frequency dictionary, sort them by length and by indices we take the first and last word)
and the last variable **average_token_length**(put all words of the frequency dictionary in one word, take its length and divide it by number of these words). and then all these variables is used as **values in dictionary of the full profile of unknown language**

File: lab_1/main.py [hunk 28]
```diff
@@ -116,35 +110,40 @@ def compare_profiles(unknown_profile: dict, profile_to_compare: dict, top_n: int
         for i in top_n_words_unknown_profile:
             if i in top_n_words_known_profile:
                 top_n_common_words.append(i)
-        intersecting_words = len(top_n_common_words) / len(top_n_words_unknown_profile)
-        return round(float(intersecting_words), 2)
+        distance = len(top_n_common_words) / len(top_n_words_unknown_profile)
+        return round(float(distance), 2)
     return None
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:115]: None check to the beginning

**Review by ValeriyaKuznetsova — COMMENTED** [review]
You missed one None check :)

**Comment by ValeriyaKuznetsova** [comment]
You've asked a question about loops. The examples of loops are `for` and `while`. You either iterate over a sequence or do smth until smth happens. A list comprehension (='генератор списка') does the same thing but it's shorter and a bit faster. So, the point is: don't write a list comprehension inside a `for` loop when they iterate over the same sequence.

Anyway, your code looks great now. I like it!

File: lab_1/main.py (hunk 28 above)
Comment (reginafeles) [reply lab_1/main.py:115]: oh, ok,i didn't noticed that. now it will be improved. thanks for cheking! and i understand what the loops mean!
